File: Populating_Next_Right_Pointers_in_Each_Node.py
# Time Complexity : O(n), where n is the number of nodes in the binary tree
# Space Complexity : O(1)
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        
        level_start = root
        
        while level_start:
            curr = level_start
            
            while curr:
                if curr.left:
                    curr.left.next = curr.right
                    if curr.next:
                        curr.right.next = curr.next.left
                curr = curr.next
            
            level_start = level_start.left
        
        return root

File: Populating Next Right Pointers in Each Node.cs
//https://leetcode.com/submissions/detail/655162770/
//Time: O(Nodes)
//Space: O(Nodes)
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}
*/

public class Solution {
    public Node Connect(Node root) {
        if(root == null) {
            return root;
        }
        
        Queue<Node> queue = new Queue<Node>();
        queue.Enqueue(root);
        root.next = null;
   
        while(queue.Count > 0) {
            int size = queue.Count;
            List<Node> list = new List<Node>();
            for(int i = 0; i < size; i++) {
                Node node = queue.Dequeue();
                Node leftNode = node.left;
                Node rightNode = node.right;
                if(leftNode !=null && rightNode != null) {
                    queue.Enqueue(node.left);
                    queue.Enqueue(node.right);
                    list.Add(leftNode);     
                    list.Add(rightNode);    
                }
            }    
                        
            for(int i = 0; i < list.Count; i++) {
                if(i == list.Count - 1) {
                    list[list.Count - 1].next = null;
                } else {
                    list[i].next = list[i+1];
                }
            }
        }
        return root;
    }
}